Some fixes for aspect ratio with padding and double-scan

carlosefr opened the pull request:
Fixes for most of the issues referenced in https://github.com/schellingb/dosbox-pure/issues/512#issuecomment-2438927259. Also, makes the choices a bit more undestandable to less advanced users.

The overscan area had a different thickness depending on whether or not a double-scanning choice was selected. I also attempted to fix this (the overscan area was already inaccurate to begin with, so this shouldn't make it worse at least).

This uncovered an issue with `vgaonly` emulation, whereas mode 13h was being handled as 320x400 instead of 320x200 with line-doubling. This seems to have been something that was left over after some other underlying issue got fixed. See: https://github.com/jwilk-mirrors/dosbox/commit/74fe003a63d3ebe01d5def9d35c16c1bcc8dd93f

It also uncovered an issue with double-scan being applied for EGA and CGA cards, when these never did double-scanning on real hardware (some particular EGA cards might, see [this thread on Vogons](https://www.vogons.org/viewtopic.php?t=85894), but that isn't expected in this case).

schellingb commented:
This seems overall good. Just one quick confirmation, the PR removes this (recently added) code:

```c
		float realratio = (((Bit32u)render.src.width<<(Bit32u)render.src.dblw) / (((Bit32u)render.src.height<<(Bit32u)render.src.dblh) * (float)render.src.ratio));
		float ratio = (float)buf.width / buf.height, diff = (ratio - realratio), ratioq = ratio * 0.25f;
		buf.ratio = ((diff < -ratioq) ? (ratio * 2) : ((diff > ratioq) ? (ratio * 0.5f) : ratio));
```

What I tried to do here is
- Calculate the "real" aspect ratio
- Calculate the square pixel ratio
- If doubling the square pixel ratio is closer to the real aspect ratio than non-doubled, use that
- If halving the square pixel ratio is closer to the real aspect ratio than non-halved, use that

I don't know if there's cases where tripling the ratio would be even closer, then it would more code or a loop. Maybe we can actually come up with math instead of this brute force approach 😅

carlosefr commented:
Yes, I removed that bit and replaced with the following code:
```
buf.ratio = ((float)render.src.width / (float)render.src.height);
```

This change fixes the following two issues with aspect ratio correction disabled that I mentioned in the other thread:

- 640x350 text mode being compressed horizontally under SVGA or EGA emulation;
- 720x400 in the start menu being compressed horizontally under VGA emulation.

Unless I'm missing something subtle, `src.width/src.height` will always make pixels square (it's the definition of square pixels). The only things that would otherwise make pixels non-square are `dblw`, `dblh`, and the `src.ratio` (pixel ratio), all unused when aspect ratio correction is disabled.

I'd have to double-check this to confirm, but I'm pretty sure that simple formula would also work for the case where line-doubling is enabled but aspect ratio correction isn't (an impossible state given the choices available in the core option) as RetroArch doesn't care about the frames' dimensions when forcing it into whatever aspect ratio the core specified.

schellingb commented:
I think doubling or halving the aspect ratio still has square pixels, just every pixel gets a copy vertically or horizontally.
Showing 640x200 with an aspect ratio of 3.2 seems weird, this would change it to a more reasonable 1.6
Showing 320x480 with an aspect ratio of 0.667 also weird, this would change it to a more reasonable 1.333

0.9.9 used to just double the width if the ratio was < 1 and halve it if it was > 2. This new approach by approximating the corrected aspect ratio seems to make more sense because we see some VGA trickery like the Pinball Illusion startup sequence that actually ends up with aspect ratios less than 1.0. It's very possible that I messed up the math with `ratioq` and comparing against that. I might need to think a bit more about it :-) It could double and halve it and then just measure which one is closest but there has to be a more logical way, just need to think a bit harder :-)

carlosefr commented:
Showing 640x200 with an aspect ratio of 3.2 and 320x480 with an aspect ratio of 0.667 is **exactly** what should happen with no aspect ratio correction. It's the very definition of square pixels. I agree it doesn't look good, but that's why there's a need for an aspect ratio correction option. 🙂

Disabling aspect ratio correction and still getting an aspect ratio corrected output doesn't sound right, and the core shouldn't do that.

With aspect ratio correction enabled (any variant), any mode that I've tested displays correctly no matter if they're double-scanned, single-scanned, with or without padding to 4:3, because it all derives from the pixel ratio (calculated in `vga_draw.cpp`).

This includes Pinball Illusions: the square pixels aspect ratios are correct (albeit ulgy), and the corrected aspect ratios are correct too.

I don't feel a need for approximations, really. But I think I get where you're coming from based on what you recently said on #512. Perhaps that middle ground between square pixels and full aspect ratio correction would fit better as an additional choice?

I don't know what that option could be called. I guess it depends on what it chooses to correct vs. not correct. But I do know that the `main` branch right now has some wrong behavior both ways, and that this PR fixes that. 😉

schellingb commented:
> Disabling aspect ratio correction and still getting an aspect ratio corrected output doesn't sound right, and the core shouldn't do that.

I don't necessarily disagree but I don't want to change the behavior of how the default non-aspect ratio corrected mode works and has worked ever since the core was released. No one has ever complained about this behavior even if it might be incorrect on a technical level. So for me it's either going back to the old way of forcing an aspect ratio between 1 and 2, or approximating the corrected aspect ratio by doubling or halving. Keep in mind it's always possible to set the "Aspect Ratio" option in RetroArch to "1:1 PAR" to enforce square pixels.

You are right though, the code you removed was incorrect and it needs to be rewritten. The modes you listed should not compress the image like it did. So I will merge this but will fiddle with it some more, see you in #512 :-)

Thanks for helping out, getting this right both technically and usability-wise is a bit of a balancing act, the PRs are appreciated! 🙏

carlosefr commented:
I have a suggestion for that, but I'll comment on #512. Thanks for merging. 👍
